syntax = "proto3";

option go_package = "pb/game";

import "common/empty.proto";
import "common/common.proto";
import "common/common_message.proto";
import "logic/component/scene_comp.proto";

option cc_generic_services = true;

message PlayerEnterGameNodeRequest
{
  uint64 player_id = 1;
  uint64 session_id = 2;
  uint32 centre_node_id = 3;
}

message ClientSendMessageToPlayerRequest
{
  MessageBody message_body = 1;
  uint64 session_id = 2;
}

message ClientSendMessageToPlayerResponse
{
  MessageBody message_body = 1;
  uint64 session_id = 2;
}

message SessionDisconnectRequest
{
  uint64 player_id = 1;
}

message GameNodeConnectRequest
{
  NetworkAddress rpc_client = 1;
  uint32 gate_node_id = 2;
}

message RegisterPlayerSessionRequest
{
  uint64 session_id = 1;
  uint64 player_id = 2;
}

message Centre2GsEnterSceneRequest
{
  uint64 player_id = 1;
  uint64 scene_id = 2;
}

message CreateSceneRequest
{
  uint32 config_id = 1;
}

message CreateSceneResponse
{
  SceneInfoPBComp scene_info = 1;
}

service GameService {
  rpc PlayerEnterGameNode (PlayerEnterGameNodeRequest) returns (Empty);
  rpc SendMessageToPlayer (NodeRouteMessageRequest) returns (NodeRouteMessageResponse);
  rpc ClientSendMessageToPlayer (ClientSendMessageToPlayerRequest) returns (ClientSendMessageToPlayerResponse);
  rpc SessionDisconnect (SessionDisconnectRequest) returns (Empty); 
  rpc RegisterGateNode (RegisterGateNodeRequest) returns (Empty);
  rpc CentreSendToPlayerViaGameNode (NodeRouteMessageRequest) returns (Empty);
  rpc InvokePlayerService (NodeRouteMessageRequest) returns (NodeRouteMessageResponse);
  rpc RouteNodeStringMsg (RouteMsgStringRequest) returns (RouteMsgStringResponse);
  rpc RoutePlayerStringMsg (RoutePlayerMsgStringRequest) returns (RoutePlayerMsgStringResponse);
  rpc UpdateSessionDetail (RegisterPlayerSessionRequest) returns (Empty);
  rpc EnterScene (Centre2GsEnterSceneRequest) returns (Empty);
  rpc CreateScene (CreateSceneRequest) returns (CreateSceneResponse);
}
